cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -411,26 +415,231 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
+        // if the schema is not known, generate it from the provided output schema and the selected fields
+        rightSchema = deriveInputSchema(stageName, rightName, key, keySchema,
+                                        joinDefinition.getSelectedFields(), joinDefinition.getOutputSchema());
+      }
+
+      // drop fields that aren't included in the final output
+      Set<String> requiredFields = new HashSet<>(key);
+      for (JoinField joinField : joinDefinition.getSelectedFields()) {
+        if (!joinField.getStageName().equals(rightName)) {
+          continue;
+        }
+        requiredFields.add(joinField.getFieldName());
+      }
+      rightSchema = trimSchema(rightSchema, requiredFields);
+
       // JoinCollection contains the stage name,  SparkCollection, schema, joinkey,
       // whether it's required, and whether to broadcast
-      toJoin.add(new JoinCollection(rightName, inputDataCollections.get(rightName),
-                                    right.getSchema(), stageKeys.get(rightName),
+      toJoin.add(new JoinCollection(rightName, inputDataCollections.get(rightName), rightSchema, key,
                                     right.isRequired(), right.isBroadcast()));
     }
 
+    String leftName = left.getStageName();
+    List<String> leftKey = stageKeys.get(left.getStageName());
+    if (leftSchema == null) {
+      if (keySchema == null) {
+        keySchema = deriveKeySchema(stageName, stageKeys, joinDefinition);
+      }
+      leftSchema = deriveInputSchema(stageName, leftName, leftKey, keySchema,
+                                     joinDefinition.getSelectedFields(), joinDefinition.getOutputSchema());
+    }
+
     // JoinRequest contains the left side of the join, plus 1 or more other stages to join to.
-    JoinRequest joinRequest = new JoinRequest(left.getStageName(), stageKeys.get(left.getStageName()), leftSchema,
+    JoinRequest joinRequest = new JoinRequest(leftName, leftKey, leftSchema,
                                               left.isRequired(), onKeys.isNullSafe(),
                                               joinDefinition.getSelectedFields(),
                                               joinDefinition.getOutputSchema(), toJoin);
     return leftCollection.join(joinRequest);
   }
 
+  /**
+   * Derive the key schema based on the provided output schema and the final list of selected fields.
+   * This is not possible if the key is not present in the output schema, in which case the user will need to
+   * add it to the output schema. However, this is an unlikely scenario as most joins will want to
+   * preserve the key.
+   */
+  @VisibleForTesting
+  static List<Schema> deriveKeySchema(String joinerStageName, Map<String, List<String>> keys,
+                                      JoinDefinition joinDefinition) {
+    int numKeyFields = keys.values().iterator().next().size();
+
+    // (stage, field) -> JoinField
+    Table<String, String, JoinField> fieldTable = HashBasedTable.create();
```

Comment:
Just curious - is there any benefit to use `Table` instead of `Map<String, Map<String, JoinField>>`?